lib/blocktree/blocktree.go

```diff
@@ -157,10 +157,6 @@ func (bt *BlockTree) Prune(newRoot Hash) (pruned []Hash) {
 	// get pruned nodes
 	pruned = bt.head.getAllDescendantsExcluding(nil, newRoot)
 
-	// set blocktree with new root node
-	next := newBlockTreeFromNode(n, bt.db)
-	*bt = *next
-
```

Comment:
we actually still need to prune the blocktree at the end of `SetFinalizedHash`, so we'll have to re-add this. I think this could be done by first copying the original blocktree (will need to implement a `Copy` func) and using that to check `IsDescendantOf`

Reply:
I've implemented DeepCopy for BlockTree and for node, but I don't understand where/how it should be used. Please let me know how these should be used.

Reply:
the current code is this, the issue is that `bt.prune` deletes the previous nodes, causing `IsDescendantOf` to fail
```
	pruned := bs.bt.Prune(hash)
	for _, rem := range pruned {
		header, err := bs.GetHeader(rem)
		if err != nil {
			return err
		}

		if des, _ := bs.IsDescendantOf(rem, hash); des {
			continue
		}

		err = bs.DeleteBlock(rem)
		if err != nil {
			return err
		}

		bs.pruneKeyCh <- header
	}
```

instead you could copy the blocktree before pruning, then use the copy to check `IsDescendantOf`
```
        copy := bs.bt.DeepCopy()
	pruned := bs.bt.Prune(hash)
	for _, rem := range pruned {
		header, err := bs.GetHeader(rem)
		if err != nil {
			return err
		}

		if des, _ := copy.IsDescendantOf(rem, hash); des {
			continue
		}

		err = bs.DeleteBlock(rem)
		if err != nil {
			return err
		}

		bs.pruneKeyCh <- header
	}
```